### ISSUE TITLE
integrate literature search into app

### ISSUE THREAD

**Issue by zfisch** [BRCAChallenge/brca-exchange#571]

**Comment by melissacline** [comment]
Chloe is working on testing literature searching, but set it aside to focus on the priors.

**Comment by rcurrie** [comment]
@zfisch @melissacline  I have a sample end to end generated [literature.json.zip](https://github.com/BRCAChallenge/brca-exchange/files/2488650/literature.json.zip). This just does the simple hgvsCoding match to get started. I haven't come up the curve of running a full sandbox though so can't ingest and test within the site. My sense is I should focus on the crawler and generation of this file vs. the site, but also happy to come up the site curve. For the former can you try the file to verify the format and content 'works' in the site against joe's pull request? 

By inspection I think there is one issue with the current pull request which looks for ['geneSnippets'](https://github.com/BRCAChallenge/brca-exchange/pull/870/files#diff-81108e9067e053af24a90b086f08c938R39) but the pipeline output has nothing in this with all the text snippets in 'mutSnippets' @strbean ?

**Comment by zfisch** [comment]
Awesome! I'll take a look right now.

**Comment by zfisch** [comment]
I have a version now up on my sandbox: http://52.229.50.68:8008/variant/144434

I noticed the variants in literature.json are coming from the Genomic_Coordinate_hg38 column -- they should be coming from pyhgvs_Genomic_Coordinate_hg38. As a result, some variants are not showing literature search results because of differences between the values.

**Comment by zfisch** [comment]
A simple example:

The first variant listed in literature.json, `chr13:32363529:T>G`, is represented as `chr13:g.32363529:T>G` in the variant table of the database (notice the added `g.`). When the request is made to the back end looking for literature search results for the variant `chr13:g.32363529:T>G`, nothing is found -- see http://52.229.50.68:8008/variant/147916.

There are other normalizations besides adding the `g.` that could have been made when converting Genomic_Coordinate_hg38 to pyhgvs_Genomic_Coordinate_hg38.

If you look in built_with_change_types.tsv, you'll find that `chr13:32363529:T>G` is the Genomic_Coordinate_hg38 value, and `chr13:g.32363529:T>G` is the pyhgvs_Genomic_Coordinate_hg38 value. At some point in the litsearch pipeline, it looks like we're searching for Genomic_Coordinate_hg38 values in the papers (or at least using them as keys in the literature.json output), when we should be searching for pyhgvs_Genomic_Coordinate_hg38 values.

**Comment by zfisch** [comment]
To clarify, the primary key in each built file is effectively "pyhgvs_Genomic_Coordinate_hg38", not "pyhgvs_cDNA" or any of the other pyhgvs values. To make the current deployment setup work properly, the keys for the variants in the literature.json output should be the pyhgvs_Genomic_Coordinate_hg38 values for the variant in built.

You may be able to solve the issue by changing:
```
built_pyhgvs[variant["pyhgvs_cDNA"]] = variant["Genomic_Coordinate_hg38"]
# strip "uncertainty" parens from protein nomenclature
built_pyhgvs[variant["pyhgvs_Protein"].replace('(', '').replace(')','')] = variant["Genomic_Coordinate_hg38"]
```
to:
```
built_pyhgvs[variant["pyhgvs_cDNA"]] = variant["pyhgvs_Genomic_Coordinate_hg38"]
# strip "uncertainty" parens from protein nomenclature
built_pyhgvs[variant["pyhgvs_Protein"].replace('(', '').replace(')','')] = variant["pyhgvs_Genomic_Coordinate_hg38"]
```

I have a dr. appointment early in the day today, but I'll check back in the afternoon to see if a call would be helpful 👍

**Comment by zfisch** [comment]
Sounds good to me!

**Comment by rcurrie** [comment]
Bit further down the HGVS rabbit hole... Tinkering with biocommons hgvs package which appears to be significantly more extensive and up to date. Includes a dockerized version that downloads all the references (~10G) and exposes it as a postgres db (also available as a service from invitae) so we can run locally an fast (for the time being using the hosted version). But it's also a bit more particular it appears then pyhgvs, or maybe its just more standardized:

chr17:g.43127865:A>G is not accepted but chr17:g.43127865A>G is
chr17:g.43127866:GAC>G is not accepted but chr17:g.43127866delGACinsG is

Assuming its useful to explore a bit this towards getting more up to date with HGVS?

Related to this from the output of the crawler I think the only 'safe' variants to use are ones with the transcript specified as others like genomic are ambiguous unless we know the reference (37 vs. 38). So my new rough plan:

1) Convert pyhgvs_Genomic_Coordinate_hg38 into canonical HG38 genomic variants using the new hgvs package towards this being the stable primary identifier.

2) Try and parse any variant found by the crawler that starts with "NM" using the hgvs package (many will not work, for example lots of NM_000059.3:c.6172T>None or NM_007294.3:c.1611A>AA)

3) Match (at this point its very reliable as both are in canonical HGVS

@zfisch @melissacline

**Comment by izcram** [comment]
@rcurrie FYI: biocommons docker is already partially integrated in the pipeline:
 https://github.com/BRCAChallenge/brca-exchange/blob/master/pipeline/Makefile#L33
A small part is missing though: https://github.com/BRCAChallenge/brca-exchange/issues/917

I also made a first small step towards a reusable component for common transformations: https://github.com/BRCAChallenge/brca-exchange/pull/915/files#diff-11731aaa37593efc53a0fa649922cea5

**Comment by rcurrie** [comment]
@izcram I have UTA and Seqrepo running locally now via Makefile. UTA didn't make that big a difference but Seqrepo dropped runtime for a single paper from 3.5 minutes to 30 seconds. Not all tied up in a nice bow but the basic ceremony and environment variables is [here](https://github.com/BRCAChallenge/pubMunch-BRCA/blob/docker/docker/Makefile#L21)

**Comment by rcurrie** [comment]
Our current primary key via a [comment by reece](https://github.com/biocommons/hgvs/issues/434) isn't HGVS and has a babelfish comment that mentions brcaexchange and that this type of variant he's loathe to parse as its [under specified](https://github.com/biocommons/hgvs/issues/447). Spelunking a bit to sort this out with the library...

**Comment by zfisch** [comment]
Interesting -- sounds like a conversion between normalized hgvs and brcaexchange genomic coordinate isn't theoretically difficult, and would solve a lot of problems (on our end at least). I've never spoken with Reece directly, but do we have an open channel to discuss?

**Comment by rcurrie** [comment]
Having trouble with anything other than simple substitutions. As an example:

BRCA Exchange Key: chr17:g.43127866:GAC>G
parser.parse_hgvs_variant("chr17:g.43127866delGACinsG") => chr17:g.43127866delinsG

so the library isn't doing the conversion, as I understand it to:
chr17:g.43127866_43127868delinsG

Also when mapping a coding hgvs to genomic the library uses:
NC_000017.11:g.43057117C>T instead of chr17:g.43057117C>T

Easy to hard code chr17 -> NC_000017.11 for now but just a few observations that makes my first approach of letting the library handle it all without 'hacks' less likely. No action at this point, want to get a new literature.json out with just single subs.

**Comment by rcurrie** [comment]
@zfisch Here's an updated [literature.json.zip](https://github.com/BRCAChallenge/brca-exchange/files/2508906/literature.json.zip). The articles key should be identical to the existing so no new parsing there. The variants list is cleaned up and pruned a bit:

```
    "variants": {
        "chr13:g.32326282:G>A": {    # key is pyhgvs_Genomic_Coordinate_38
            "Alt": "A",
            "Chr": 13,
            "Pos": 32326282,
            "Ref": "G",
            "norm_c_hgvs": "NM_000059.3:c.516G>A",  # hgvs library parsed
            "norm_g_hgvs": "NC_000013.11:g.32326282G>A",  # hgvs library parsed and mapped
            "pmid": "23199084",  # used to be docId - now matches articles
            "pyhgvs_Genomic_Coordinate_38": "chr13:g.32326282:G>A",
            "pyhgvs_cDNA": "NM_000059.3:c.516G>A",
            # Just snippet now instead of all the different flavors
            "snippet": "nd six major recurrent founder mutations (three BRCA1 c.212+3A> (BIC: IVS5+3A>G), c.2359dupG (BIC: 2478insG), c.3661G>T (BIC: 3780G>T) and three BRCA2<<< c.516+1G>A>>> (BIC: IVS6+1G>A), c.6275_6276delTT (BIC: 6503_6504delTT), c.8904delC (BIC: 9132delC) alterations), which accounted for nearly 60% of all mutations id|BRCA1 c.2359dupG and BRCA2<<< c.516+1G>A>>> .....
```

Take a look at let me know what you think - I have more debugging and validation to do and if its a pain to adjust the import scripts I can easily switch back to the old names (docId, geneSnippets etc...)

**Comment by zfisch** [comment]
Nice! Much cleaner. I just pushed code to integrate this new format and updated http://52.229.50.68:8008/variant/147916 -- looking good.

**Comment by rcurrie** [comment]
Excellent!!!! Thanks for adjusting and glad it looks better. I have some additional debugging to do with this library but will return to the crawler portion to see about getting a larger set of ~100 reliably.

**Comment by rcurrie** [comment]
@strbean I'm getting an error from a missing .hg.conf, but I've downloaded all the references locally. Does pubMunch still need to hit a browser server? (This is in the find phase)
```
  File "/pubMunch/lib/geneFinder.py", line 712, in findGenesResolveByType
    seqDict            = resolveSeqs(markers["dnaSeq"], seqCache)
  File "/pubMunch/lib/geneFinder.py", line 474, in resolveSeqs
    blatClient = seqMapLocal.BlatClient(pubConf.genomeDataDir, ["hg19"])
  File "/pubMunch/lib/seqMapLocal.py", line 34, in __init__
    self._getBlatServersFromHgCentral()
  File "/pubMunch/lib/seqMapLocal.py", line 37, in _getBlatServersFromHgCentral
    config = maxCommon.parseConfig("~/.hg.conf")
  File "/pubMunch/lib/maxCommon.py", line 510, in parseConfig
    f = open(os.path.expanduser(f))
IOError: [Errno 2] No such file or directory: '/.hg.conf'
Makefile:63: recipe for target 'find' failed
make: *** [find] Error 1
```

**Comment by rcurrie** [comment]
@strbean Sorted it out - missing commit from Audrey to turn off DNA search.

**Comment by rcurrie** [comment]
I have everything running end to end now and have kicked off a crawl of 500 random PMID's to see how long it takes as well as get a sense of how many we are actually able to download. Will also help to get a broader idea of how accurately all the moving pieces are working.

**Comment by rcurrie** [comment]
500 papers crawled with 396 successfully downloaded and converted in 129 minutes

Total # papers in pubmed with BRCA in title: 16908

16908 * 129 minutes / 500 papers  / 60 minutes = 72 hours to download and convert

**Comment by zfisch** [comment]
Wow! from 2ish weeks to 72 hours 👏

**Comment by zfisch** [comment]
@rcurrie what final steps do you feel are necessary on the pubmunch side before we can get to a v1? I'll share my sandbox with the team and resolve any outstanding issues in the meantime.

**Comment by rcurrie** [comment]
[literature.json.zip](https://github.com/BRCAChallenge/brca-exchange/files/2520892/literature.json.zip)
@zfisch Here's the output from a crawl of 500 papers (396 successfully). It doesn't include the paper I sent before (23199084) so may be a good test on what the right thing to do ingestion wise is. I haven't delved yet into the crawler and how it handles papers its already seen - whether we can and should re-find as the finding logic improves etc... In any case passing on in case you want to load up a larger set for @melissacline @amycoffin to smoke test. I'm going to go through the logs of the crawl to see if there are any noticeable errors I should deal with but in parallel if there are any data related changes I can make those (for example how the snippets come come out etc...).

**Comment by rcurrie** [comment]
Short term steps to just get a crawl that we can run (but not a tidy bundled up reproducible docker with test cases etc...):

- [ ] Go through the logs to see if any big errors are getting thrown, ignoring if say 10-20% of the papers are failing to download
- [ ] Have various people take a look at the UI and content to smoke test wise see if the variant finding code is doing the right thing - not insanely high false positives etc...
- [ ] Try loading successive literature.json's verify the migrations do the right thing (whatever that may be) so that if we go into production we can upgrade
- [ ] Do a bigger crawl - maybe ~1k-2k and load up

At that point I could just run myself for 72 hours. Its running in a docker but I haven't put all the nice things like help, one single run.py, auto download of references (currently at 25G). All that I could do after the above, or we could just sandbox a small run for a while and not put into production until the docker and pipeline is production ready. I'm inclined to the former given the nature of this as a 'please point me to less then infinity and more then zero' bar :-)

**Comment by rcurrie** [comment]
@melissacline @zfisch @amycoffin The direct list of mutations and snippets found by the pubMunch crawler for [this paper](https://www.ncbi.nlm.nih.gov/pubmed/23199084) is [here](https://docs.google.com/spreadsheets/d/1uI1R2ShvxjTTsuNsPi_dnigBrIubUWKUA7AxvJVCL4Y/edit?usp=sharing). I've put the hgvs that the crawler finds next to the snippets it generates. [correlate.py](https://github.com/BRCAChallenge/pubMunch-BRCA/blob/docker/docker/correlate.py) take this, tries to parse and map each hgvs string and if successful adds to literature.json that is ingested to the site. The simple thing might be to get rid of the <<< >>> and just have the javascript look for a match of c.XXX from the coding hgvs I emit (post parse/map) as for some the transcript name differs.

Example

NM_007294.3:c.5074G>A|NM_007297.3:c.5074G>A

population, 8.5% of consecutive unselected breast cancer cases, 7.9% of ovarian cancer cases and 40% of male breast cancer cases [6, 146].  The BRCA1<<< c.5074G>A>>> (BIC: 5193G>A) is another founder mutation recently identified in Icelandic population, however it is extremely rare and contributes only to 1% of br

**Comment by melissacline** [comment]
@Rob Currie I see a couple issues with the second output row.  The text snippet begins, “One BRCA2 mutation c.4258delG (BIC:<<< 4486delG>>>)…“.  The mapping in Column 1 is “NM_007300.3:c.4486G>None|NM_007297.3:c.4486G>None”.    

1. We've picked up the wrong gene.  Both NM_007300.3 and NM_007297.3 are BRCA1 transcripts, not BRCA2 (though neither is the canonical BRCA1 transcript NM_007294.3).  The geneSymbol column in the spreadsheet also indicates that it was mapped to BRCA1.

2. We picked up the wrong position.  It looks like the position of the mutation has been picked up by the BIC identifier (4486delG), where the correct HGVS nucleotide identifier is c.4258delG.  BIC identifiers often contain a different numerical position than HGVS identifiers, which comes from a different start position (I believe that BIC starts at the CDS start, while HGVS starts at the start codon).

**Comment by melissacline** [comment]
3. In Row 13, we correctly identified a BRCA2 variant, but didn't identify the correct position.  The text snippet begins "The most prevalent BRCA2 mutations are c.8363G>A (BIC: 8591G>A/W2788X),<<< c.8754+1G>A>>> (BIC: IVS21-1G>A)" and the HGVS in Column A is  
NM_000059.3:c.8754G>A.  In this case, the variant in the text is an intronic variant (c.8754+1G>A), while we've picked up the closest exonic variant (c.8754G>A).  Joe had warned me that the crawler was not picking up all of the intronic variants: for variants of the syntax c.X+Y(change) or c.X-Y(change), it was picking up X, (change) but not (always? ever?) Y.  I had told him not to go down the rabbit hole of trying to fix this right now, that we can exclude intronic variants in our first release.

**Comment by melissacline** [comment]
4. In Row 5, we picked up the correct gene and position, but not the canonical transcript.  " In the Greek population four BRCA1 mutations <<<(c.5526dupC>>>, c.5212G>A (BIC: 5331G>A/G1738R), c.5251C>T (BIC: 5370C>T/R1751X), c.5467G>A (BIC: 5586G>A), account for 54% of all mutations" is translated to  NM_007300.3:c.5526C>CC.  We somehow ended up with transcript NM_007300.3 and not the canonical transcript, NM_007294.3.  This would throw off the variant position.  According to Mutalyzer, NM_007300.3:c.5526dupC corresponds to  NM_007294.3:c.5463dupC, and NM_007294.3:c.5526dupC corresponds to NM_007300.3:c.5589dupC.

**Comment by melissacline** [comment]
4b. In Line 11, we parsed the same text snippet for a different variant, and everything went fine - yay!  "In the Greek population four BRCA1 mutations (c.5526dupC, c.5212G>A (BIC: 5331G>A/G1738R),<<< c.5251C>T>>> (BIC: 5370C>T/R1751X), c.5467G>A (BIC: 5586G>A), account for..." translated to  NM_007294.3:c.5251C>T.

**Comment by rcurrie** [comment]
I haven't cherry picked Joe's [pyhgvs commit](https://github.com/BRCAChallenge/pubMunch-BRCA/commit/dfabf12d477846bf7667470988fc22b4230d7881#diff-59f5d8ede3dabaa388a170cdcac88841) on varFinder.py which I believe is the relevant file/code as the correlation.py trys to use the updated hgvs library for matchin. At 1400 lines this file has quite a bit of heuristics to try and find things likely in a general fashion for multiple genes. We might be able to get it working more consistently while just focusing on BRCA. That said I haven't spent any time on this piece of the puzzle and don't really have a good sense of this files approach. All the other commits are there from Max/Audry et al but none of this was every put into production as far as I can tell so not sure about how much scrutiny accross other papers there is. I did another crawl of 100 random PMID's and will look through it a bit before our call tomorrow.

**Comment by melissacline** [comment]
Viewing this as a prototype method, it's fair game for us to focus on BRCA1 and BRCA2, and assume the canonical transcripts of NM_007294.3 for BRCA1 and NM_000059.3 for BRCA2.  All of this illustrates the challenges of text mining as a general problem.  The general approach seems worthy of a Ph.D. thesis project for some future grad student.  We know we're not going to reach perfect, but we can probably reach useful.  And as long as we keep showing those text snippets, we decrease the cost of the false positives.

**Comment by rcurrie** [comment]
Updated short term plan:

- [ ] Crawl 1k random PMIDs + 2 test papers (23199084 & 30186769) and add to sandbox
- [ ] Walk through all debug output and results for 23199084
- [ ] Get feedback on 1k in sandbox from users to determine whether to do full crawl or more engineering on the var finder and correlation

**Comment by rcurrie** [comment]
Crawled 1000 random PMID's + 2 test papers (23199084 & 30186769):

769 downloaded in 248 minutes & 768 converted in 35 minutes
(~80 hours for full crawl of 16,908 papers)
17093 mutations found
860 of those correlated to mutations in built

Passed the output to @zfisch for ingest into the sand box to gather impressions.

**Comment by zfisch** [comment]
The latest crawl (mentioned above) is now live at http://52.229.50.68:8008. I've attached a file that lists all the variants that have lit results from the crawl and are eligible for review.
[variants_with_lit_results.txt](https://github.com/BRCAChallenge/brca-exchange/files/2532391/variants_with_lit_results.txt)

**Comment by melissacline** [comment]
I've looked through some random hits in the crawl results, and overall I'm excited!  There were some false positives, but overall the text snippets were sufficient to weed out most false positives.   The false positives I noticed seemed to be tables where the gene name (BRCA1/2) appeared in one row, and the variant name appeared in a subsequent row (with a different gene).  I've attached my notes. 
[lit.search.sample.notes.txt](https://github.com/BRCAChallenge/brca-exchange/files/2546742/lit.search.sample.notes.txt)

**Comment by zfisch** [comment]
I noticed some "NaN" snippets too -- @rcurrie any idea what's going on there? As a more general note, it seems like we shouldn't report a match any time we can't display the snippet including the match.

**Comment by rcurrie** [comment]
Feedback from Samantha pasted here from Slack:

*chr17:g.43092805:T>A* -  I couldn't find this paper through searches in pubmed or google using the usual script (see below).  Validated that it was a real paper.

"BRCA1" ("2726A>T" | "2726A->T" | "2726A-->T" | "2726A/T" | "Asn909Ile" | "N909I")

HGMD didn't have the paper we had, but has one that this search does not - Chen (2003) Breast Cancer Res Treat 77, 133 (PMID: 12602912)

=> Rob Comment: This PMID is in the overall list that would be crawled on a comprehensive run

*chr13:g.32356590:C>G* - I couldn't find this paper through searches in pubmed or google using the usual script (see below). Validated that it was a real paper.

"BRCA2" ("7598C>G" | "7598C->G" | "7598C-->G" | "7598C/G" | "Ser2533Cys" | "S2533C")

HGMD didn't have any paper listed.


*chr13:g.32326569:G>A* - False positive paper.  PTEN paper, not BRCA2. No papers missed.

*chr17:g.43094717:C>A*- This is an interesting scenario. It is a somatic change, not a germline case, so I think it should have come up but would have been ruled out. Google search found this paper, but pubmed did not using script below.

"BRCA1" ("814G>T" | "814G->T" | "814G-->T" | "814G/T" | "Glu272*" | "Glu272X" | "Glu272Ter" | "E272*" | "E272X")

Nothing in HGMD
*chr17:g.43104882:T>C* - False positive paper.  RAD51C paper, not BRCA2.  No papers missed.

**Comment by rcurrie** [comment]
Started full download of 16980 papers with BRCA in the title or abstract from pubmed.

**Comment by rcurrie** [comment]
Download complete:
16980 papers attempted, 14088 papers successfully downloaded (83%) over 116 hours consuming around 20G of disk space. Moving on to convert, find and correlate.

**Comment by rcurrie** [comment]
@zfisch @falquaddoomi Initial output from full crawl is available here:

http://public.gi.ucsc.edu/~rcurrie/literature-25-11-2018.json

# Papers: 14088
# Variants: 3754

Seems really low yield but though I'd circulate to see if it ingests into the sandbox.

**Comment by rcurrie** [comment]
As per call today here's the short term plan:

- [ ] @zfisch Put current file up in a sandbox with the new UI and @amycoffin get some impression level feedback from users. If its useful and the feedback is 'we'd rather have this as it is now vs. better later' then move forward with putting this into production.
- [ ] @rcurrie Trace through a few papers that yielded zero variants to see where the filtering is. Likely in correlate but want to be sure.
- [ ] @rcurrie Code up a much simpler variant finder + correlate to see how it performs

**Comment by zfisch** [comment]
New lit search data is up at http://52.229.50.68:8008 and ready for review.

Note: Some papers are coming to us without a year associated with them, namely pmids: 27869446, 15258337, 10464643, and 10464641. @rcurrie is this an issue happening on the backend or is there a valid reason these papers have no year associated with them?

**Comment by zfisch** [comment]
@amycoffin could you circulate this latest data for feedback?

**Comment by falquaddoomi** [comment]
@amycoffin Yes, that's the most current version of the UI.

**Comment by zfisch** [comment]
Let's let @melissacline take a look and if she's ok with it, let's deploy it. We can discuss whether or not to include a beta tag as well.

In the meantime:

@rcurrie in the current deployment implementation, we add papers we don't already have in the db, then associate every variant found in literature.json with an existing paper and add it to the db. I plan to tweak this so we don't end up with duplicate correlations, but also want to make sure it's able to update false positives as we make improvements. **does the current correlation implementation check every variant across every paper every time we run lit search? i.e. does the literature.json file always contain a complete picture of all literature results for all variants in brcaexchange?** if so, I plan to soft delete all correlations in the db that aren't in the latest literature.json results with each deployment.

**Comment by amycoffin** [comment]
@zfisch @rcurrie The GC meetings are happening next week, 12/3-12/7, and I think that we can expect a decision about whether LitSearch is good to go to beta either late Friday 12/7 or early Monday 12/10. I'd like to clarify that we are currently **good to circulate the sandbox for feedback, but I would still like the chance to double check the overall UI and Docs.** There might be one or two small tweaks I submit as issues, that have been lingering since before the holiday. 

@melissacline, IMHO the decision about beta will basically hinge on whether you think the UI/Docs are in good shape and have gone through enough EGG vetting, and whether GC feedback is substantial enough to wait one more month.

**Comment by rcurrie** [comment]
@zfisch wrt dates it appears for those papers medlineCreatedDate is an empty string. But, for example, 'medlineCompletedDate': u'2018-04-23' is present - but I'm not sure its reasonable to switch to it completely. I'll do a bit of exploration in the papers db and get back to this issue, for its on the crawler's end.

**Comment by rcurrie** [comment]
@zfisch The current implementation does correlate across all variants found so doing a soft delete at this point is probably the right thing to do i.e. assume literature.json replaces everything. I have not looked closely at how pubmunch handles incremental downloads (ie a few new papers were found and downloaded, are the added to the source mutations or a new file generated). I'm going to spelunk a bit into that though as part of that exploration on an alternative simpler approach. TL;DR; replace...

**Comment by rcurrie** [comment]
@zfisch Confirmed that those are the only four articles with missing year by looking at the original pubmed articles.db. Capturing in https://github.com/BRCAChallenge/pubMunch-BRCA/issues/7

**Comment by rcurrie** [comment]
@amycoffin @zfisch Something definitely wrong with my join of articles and mentions. Will have a new literature.json soon. Here's some counts by variant/pmid:

norm_g_hgvs                  pmid    
NC_000001.11:g.100149376C>T  29868112    1
NC_000001.11:g.100149388del  29868112    1
NC_000001.11:g.100149427A>T  29868112    2
NC_000001.11:g.100152418G>A  29868112    4
NC_000001.11:g.100152424del  29868112    4
NC_000001.11:g.100155115G>A  29868112    4
NC_000001.11:g.100155135T>C  29868112    4
NC_000001.11:g.100155152del  29868112    4
NC_000001.11:g.100155160T>C  29868112    4
NC_000001.11:g.100155186G>C  29868112    4
NC_000001.11:g.100155190G>A  29868112    4
NC_000001.11:g.100156196T>G  29868112    8
NC_000001.11:g.100156218C>G  29868112    4
NC_000001.11:g.100156290G>A  29868112    4
NC_000001.11:g.100156302T>A  29868112    4
NC_000001.11:g.100158288G>C  29868112    4
NC_000001.11:g.100158290C>A  29868112    4
NC_000001.11:g.100158308T>C  29868112    8
NC_000001.11:g.100158350G>A  29868112    8
NC_000001.11:g.100159259C>T  29868112    4
NC_000001.11:g.100159340G>A  29868112    4
NC_000001.11:g.100159366del  29868112    4
NC_000001.11:g.100159376C>T  29868112    4
NC_000001.11:g.100159378T>C  29868112    4
NC_000001.11:g.100159382G>T  29868112    8

**Comment by rcurrie** [comment]
Aha! I see what's going on - its an an unintended consequence of the export format that I should have picked up. We have two dictionaries: papers and variants with the latter keyed by genomic hgvs so there is no possibility of more than one mention per variant listed.

@amycoffin the actual numbers are 55657 mentions that correspond to variants in BRCA Exchange extracted from 2227 papers. Here's a sample of counts by variant by paper # mentions: 


```
norm_g_hgvs                 pmid    
NC_000017.11:g.43124095A>G  27125668     4
                            28176296     4
                            28223274     4
                            29684080     4
NC_000017.11:g.43124096T>C  11802209     8
                            12181777     4
                            12827452     8
                            12955082     4
                            15980987     4
                            18034184     8
                            20180971     8
                            21120943     8
                            21170264     4
                            21559243     8
                            22006311     8
                            22933969     8
                            24302565    16
                            24312913     8
                            24884479     8
                            25639900     8
                            28272917     8
                            28324225     8
                            28475402    16
                            29339979     8
                            29907814     4
                            30103829    16
NC_000017.11:g.43124096T>G  10811497     4
                            23199084     8
                            24309564     4
                            25639900    16
```

@melissacline @falquaddoomi We're going to need to re-think the GUI with this many hits per variant and paper.

@zfisch We need to establish a different structure for the mentions part of the exported json. Simple option would be to have each variant be a list of mentions:

```
"variants": {
 'NC_000013.11:g.32316462T>G': [{'norm_c_hgvs': 'NM_000059.3:c.2T>G',
   'pmid': 25072261,
   'snippet': 'T  1  BRCA2  6589delAG  1  BRCA2  c.8332-?-8487 þ ?del, exon 19 deleted  1  BRCA2  3393delC  1  BRCA2  8765delAG  1  BRCA2  3967delA  1  BRCA2  IVS7 þ<<< 2T4G>>>  1  BRCA2  4003G4T  1  BRCA2  7908T4A  1  BRCA2  9118-1G4C  1  BRCA2  2041delA  1  BRCA2  262_263delCT  1  BRCA2  2957_2958insG  1  BRCA2  5722_5723d'}],
 'NC_000013.11:g.32316463G>A': [{'norm_c_hgvs': 'NM_000059.3:c.3G>A',
   'pmid': 29422015,
   'snippet': 'hin the Hereditary Prostate Cancer 1 (HPC1) linkage peak on chromosome 1q on the basis of evidence that two inactivating mutations in the gene, RNASEL<<<:p.Met1Ile>>> and RNASEL:p.Glu265*, segregated with prostate cancer in chromosome 1q–linked pedigrees [15].'},
  {'norm_c_hgvs': 'NM_000059.3:c.3G>A',
   'pmid': 29422015,
   'snippet': 'hin the Hereditary Prostate Cancer 1 (HPC1) linkage peak on chromosome 1q on the basis of evidence that two inactivating mutations in the gene, RNASEL<<<:p.Met1Ile>>> and RNASEL:p.Glu265*, segregated with prostate cancer in chromosome 1q–linked pedigrees [15].'},
  {'norm_c_hgvs': 'NM_000059.3:c.3G>A',
   'pmid': 22006311,
   'snippet': 'BRCA1<<< p.M1I>>>, p.M1V, and p.C61G and TP53 p.Y126D, p.R175C, and pR175H were included as loss- of-function alleles on the basis of previous experimental results (Ta'},
  {'norm_c_hgvs': 'NM_000059.3:c.3G>A',
   'pmid': 22006311,
   'snippet': 'BRCA1<<< p.M1I>>>, p.M1V, and p.C61G and TP53 p.Y126D, p.R175C, and pR175H were included as loss-of-function alleles on the basis of previous experimental results (Table S1).'}],
 'NC_000013.11:g.32316463G>T': [{'norm_c_hgvs': 'NM_000059.3:c.3G>T',
          ....
```
Or could sub group by PMID. Whatever is easiest for ingest I suppose.

**Comment by rcurrie** [comment]
Max # Mentions for a variant: 38
Mean: 1.9
There appear to be duplicate mentions which we didn't see before due to the export problem. Going to see about filtering...

**Comment by rcurrie** [comment]
I've generated a preliminary new export json with the "variants" key consisting of a dictionary by variant and in turn a dictionary of pmids:

http://public.gi.ucsc.edu/~rcurrie/literature-03-12-2018.json

As an example of a variant with multiple papers each of which has multiple variants:

```
        "NC_000013.11:g.32316470G>T": {
            "17453335": [
                {
                    "norm_c_hgvs": "NM_000059.3:c.10G>T",
                    "pmid": 17453335,
                    "snippet": "Two novel BRCA2 germline mutations <<<(G4X>>> and 3783del10) are reported here for the first time.|Each of the five BRCA2 mutations is observ
ed only once; three are frameshift (3783del10, 4637delTA, 5950delCT), one is nonsense <<<(G4X>>>), and one is a single base substitution that results in aberrant splicing (7
235G\u00a0>\u00a0A, skipping of exons 12\u201313) [15].|h.gov/ Intramural_research/Lab_transfer/Bic/Member/ BRCA1_mutation_database.html). The two novel BRCA2 mutations desc
ribed here for the first time are<<< G4X>>> and 3783del10, located in exons 2 and 11, respectively.|829     One case of male BrCa, one case of CRC, one case of unknown prima
ry ca      335     5 (37 t+ , 43, 56 p,t+ , 57, 77)    \u00a0  \u00a0   BRCA2/exon2    <<< G4X>>>*     ter 4         Families with three cases     393     1 (42 p,t+ )     2
 (45, 55)    \u00a0   BRCA1/exon11     3819del5     ter 1242         287     2 |xon24     \u2013     \u2013     \u0394 82651-87079#     \u0394 exon24     del4429ins5Ex24g
  R     R     \u2013     2      BRCA2/exon2     238     4     G to \u03a4     ter 4    <<< G4X>>>*     N     N     \u2013     1      BRCA2/exon10     2024     599     del TT
TAT     Stop 599     2024del5b,c     F     F     12     2      BRCA2/exon11    |f deleterious alterations of the BRCA1 gene in the Greek population (11%, 5/46).  Apart from
the two novel mutations reported here for the first time <<<(G4X>>> and 3783del10 in BRCA2), other novel mutations confined to the Greek population are 3099delT and 3277insG
 in BRCA1 [11], and 3058delA in BRCA2 [12]. "
                },
                {
                    "norm_c_hgvs": "NM_000059.3:c.10G>T",
                    "pmid": 17453335,
                    "snippet": "Two novel BRCA2 germline mutations <<<(G4X>>> and 3783del10) are reported here for the first time.|Each of the five BRCA2 mutations is observ
ed only once; three are frameshift (3783del10, 4637delTA, 5950delCT), one is nonsense <<<(G4X>>>), and one is a single base substitution that results in aberrant splicing (7
235G > A, skipping of exons 12\u201313) [15].|The two novel BRCA2 mutations described here for the first time are<<< G4X>>> and 3783del10, located in exons 2 and 11, respect
ively.|u n cl as si fi ed v ar ia n t 438 Breast Cancer Res Treat (2008) 107:431\u2013441 123 Apart from the two novel mutations reported here for the first time <<<(G4X>>>
and 3783del10 in BRCA2), other novel mutations confined to the Greek population are 3099delT and 3277insG in BRCA1 [11], and 3058delA in BRCA2 [12]. "
                }
            ],
            "24010542": [
                {
                    "norm_c_hgvs": "NM_000059.3:c.10G>T",
                    "pmid": 24010542,
                    "snippet": "_5797delTA  11B  c.4409_4410delTA c.4415_4418delAGAA c.4767delA  11A  10  c.3555_3564del10 c.3919delG  9  p.Ser780X  7  c.2808_2811delAAAC c.2830delA <<< p.Gly4X>>>  5  21 23 20 22  8  c.1796_1800del5  3  1  6  c.658_659delGT  4  2  Fig. 1."
                }
            ],
```

@zfisch This may be harder to import though so the simpler alternative is to just have a flattened list keyed by variant i.e. each variant will have an array of dictionaries and its up to the server or client to organize by pmid. Either way we need to switch the

**Comment by rob** [comment]
Unsubscribe from cat facts.

**Comment by rcurrie** [comment]
Zack and I just chatted and decided to just switch the "variants" key in the json to an array.

@zfisch I've updated http://public.gi.ucsc.edu/~rcurrie/literature-03-12-2018.json

```
{
    "papers": {
        "10070866": {
            "abstract": "Reduced expression of BRCA1 has been implicated in sporadic breast cancer, although the mechanisms underlying this phenomenon remain unclear. To determine whether regulatory mutations could account for the reduced expression, we screened the promoter region by sequencing in 20 patients with sporadic disease. No mutations were detected; however, a new polymorphism consisting of a C-to-G base change within the beta-promoter was identified, with the frequency of the G allele being 0.34. Close to complete linkage disequilibrium was found between this marker and the Pro871 Leu polymorphism, situated in exon 11, which has previously been shown not to be associated with breast or ovarian cancer. This indicates that the C/G polymorphism is also unlikely to play a role in either disease. However, the strength of linkage disequilibrium between these markers permitted their use for rapid screening for genomic deletions within BRCA1. A series of 214 cases with familial breast cancer were analysed using this approach; 88/214 were heterozygous for the promoter polymorphism, thereby excluding a deletion in this region. Among the remaining patients, one hemizygous case reflecting a promoter deletion was successfully identified. Therefore, this study indicates that deletions within the beta-promoter region of BRCA1 are an uncommon event in familial breast cancer. Furthermore, it suggests that mutations within the BRCA1 promoter are unlikely to account for the reported decreased expression of BRCA1 in sporadic disease.",
            "articleId": "5010070866",
            "articleSection": "",
            .....
        }
    },
    "variants": {
       "NC_000013.11:g.32316470G>T": [
            {
                "norm_c_hgvs": "NM_000059.3:c.10G>T",
                "pmid": 28751759,
                "snippet": "65Leu) VUS, which was also present in this sample.  Laboratory 16, using an HRM pre-screening strategy, failed to identify the BRCA2 nonsense mutation<<< c.10G>T>>>; p.(Gly4*) in one sample."
            },
            {
                "norm_c_hgvs": "NM_000059.3:c.10G>T",
                "pmid": 28751759,
                "snippet": "2Glnfs*5) NA 40 bp deletion (frameshift) 2 BRCA1 (NM_007294.3); BRCA2 (NM_000059.3) NA NA Control sample\u2013no pathogenic mutations 3 BRCA2 (NM_000059.3)<<< c.10G4T>>>; p.(Gly4*) NA Point mutation (nonsense) 4 BRCA1 (NM_007294.3) c.442-?_547\u00fe ?del, NA LGR: deletion (exon 8) 5 BRCA2 (NM_000059.3) c.5909C4A; p.(Ser197|465Leu) VUS, which was also present in this sample. Laboratory 16, using an HRM pre-screening strategy, failed to identify the BRCA2 nonsense mutation<<< c.10G4T>>>; p.(Gly4*) in one sample."
            },
            {
                "norm_c_hgvs": "NM_000059.3:c.10G>T",
                "pmid": 24312913,
                "snippet": "G, 461delTC [41]  4817A>G, 8477delAGA [41]  \u2014   Greek (South eastern Europe)  5382insC and G1738R   [92]   G5331A, 3819delGTAAA [180]   8984delG [35],<<< G4X>>> and 3783del10 [92]  5382insC and G1738R   [92], 3782del10, 4512insT [180]   Slovenia  c.1193C>A [86]   c.181T>G, c.1687C>T, 5382insC,   c.844_850dupT"
            },
            {
                "norm_c_hgvs": "NM_000059.3:c.10G>T",
                "pmid": 24312913,
                "snippet": "8142del5 [28], exons 5\u201314 del Exons 1\u201317 del 5382insC [175]  c.6589delA [178, 179]  c.6589delA [178, 179]  4817A>G, 8477delAGA [41]  \u2014  8984delG [35],<<< G4X>>> and 3783del10 [92]  5382insC and G1738R [92], 3782del10, 4512insT [180]  c.5101C>T, c.5433 5436delGGAA [86] c.7806\u22122A>G, c.5291C>G, c.3975 3978dupTGC"
            },
            {
                "norm_c_hgvs": "NM_000059.3:c.10G>T",
                "pmid": 27225819,
                "snippet": "  95.96%   3.239   447.681   98.46%   BRCA2+   c.1447_1450ins4; p.Ala483AlaValfs   Path   #CQ3 TS   377.2   42.18%   227   120.384   57.25%   BRCA2+  <<< c.10G>T>>>; p.Gly4Ter   Path   #CQ5 TS   811.4   66.65%   400   173.040   89.41%   BRCA2+   c.5909C>A; p.Ser1970Ter   Path   #CQ7 TS   319.23   74.47%   105   3"
            },
            ...
```

**Comment by rcurrie** [comment]
@zfisch Sorry - of course. "variants" now keyed by "pyhgvs_Genomic_Coordinate_38" 

http://public.gi.ucsc.edu/~rcurrie/literature-03-12-2018-v2.json

```
    "variants": {
       "chr13:g.32316470:G>T": [
            {
                "Alt": "T",
                "Chr": 13,
                "Pos": 32316470,
                "Ref": "G",
                "norm_c_hgvs": "NM_000059.3:c.10G>T",
                "norm_g_hgvs": "NC_000013.11:g.32316470G>T",
                "pmid": 28751759,
                "pyhgvs_cDNA": "NM_000059.3:c.10G>T",
                "snippet": "65Leu) VUS, which was also present in this sample.  Laboratory 16, using an HRM pre-screening strategy, failed to identify the BRCA2 nonsense mut
ation<<< c.10G>T>>>; p.(Gly4*) in one sample."
            },
            {
                "Alt": "T",
                "Chr": 13,
                "Pos": 32316470,
                "Ref": "G",
                "norm_c_hgvs": "NM_000059.3:c.10G>T",
                "norm_g_hgvs": "NC_000013.11:g.32316470G>T",
                "pmid": 28751759,
                "pyhgvs_cDNA": "NM_000059.3:c.10G>T",
                "snippet": "2Glnfs*5) NA 40 bp deletion (frameshift) 2 BRCA1 (NM_007294.3); BRCA2 (NM_000059.3) NA NA Control sample\u2013no pathogenic mutations 3 BRCA2 (NM
_000059.3)<<< c.10G4T>>>; p.(Gly4*) NA Point mutation (nonsense) 4 BRCA1 (NM_007294.3) c.442-?_547\u00fe ?del, NA LGR: deletion (exon 8) 5 BRCA2 (NM_000059.3) c.5909C4A; p.(
Ser197|465Leu) VUS, which was also present in this sample. Laboratory 16, using an HRM pre-screening strategy, failed to identify the BRCA2 nonsense mutation<<< c.10G4T>>>;
p.(Gly4*) in one sample."
            },
```

**Comment by zfisch** [comment]
Nice! Curious, any reason not to include all snippets for a single variant in the same paper in a list/set?

For example, the above would instead look like:

```
"variants": {
       "chr13:g.32316470:G>T": [
            {
                "Alt": "T",
                "Chr": 13,
                "Pos": 32316470,
                "Ref": "G",
                "norm_c_hgvs": "NM_000059.3:c.10G>T",
                "norm_g_hgvs": "NC_000013.11:g.32316470G>T",
                "pmid": 28751759,
                "pyhgvs_cDNA": "NM_000059.3:c.10G>T",
                "snippet": ["65Leu) VUS, which was also present in this sample.  Laboratory 16, using an HRM pre-screening strategy, failed to identify the BRCA2 nonsense mut
ation<<< c.10G>T>>>; p.(Gly4*) in one sample.", "2Glnfs*5) NA 40 bp deletion (frameshift) 2 BRCA1 (NM_007294.3); BRCA2 (NM_000059.3) NA NA Control sample\u2013no pathogenic mutations 3 BRCA2 (NM
_000059.3)<<< c.10G4T>>>; p.(Gly4*) NA Point mutation (nonsense) 4 BRCA1 (NM_007294.3) c.442-?_547\u00fe ?del, NA LGR: deletion (exon 8) 5 BRCA2 (NM_000059.3) c.5909C4A; p.(
Ser197|465Leu) VUS, which was also present in this sample. Laboratory 16, using an HRM pre-screening strategy, failed to identify the BRCA2 nonsense mutation<<< c.10G4T>>>;
p.(Gly4*) in one sample."]
            },
```

**Comment by rcurrie** [comment]
No particular reason. Seems like it would make it cleaner on the database and UI. I'll noodle a bit more on the pandas and see if its reasonably straight forward.

**Comment by rcurrie** [comment]
@zfisch @falquaddoomi Here's an updated json where each variant has a dictionary keyed to pmid and all mentions concatenated with |:

http://public.gi.ucsc.edu/~rcurrie/literature-04-12-2018-v1.json

Questions:
- Stick with '|', another character, or [ ]
- PMID key in variants is a string although its an int - this is due to python serialization. The pmid field in the article dictionary is an int. Just a note and question if you want me to try and change it.

Sample:

```
{
    "papers": {pmid:....},
    "variants": {
        "chr13:g.32316489:C>A": {
            "25925845": "642785  3  12  21  18  25.0  54  75.0  26.0  74.0  B   c.96+41_96+56del16    intron 3  rs17878362  24  11  1  59  81.9  13  18.1  79.5  20.5  B   c.97<<<-29C>A>>>    intron 3  rs17883323  35  1  0  71  98.6  1  1.4  94.8  5.2  B   c.108G>A  p.Pro36Pro  exon 4  rs1800370  34  2  0  70  97.2  2  2.8  98.5  1.5  B| rs1642785  3  12  21  18  25.0  54  75.0  26.0  74.0  B  c.96+41_96+56del16  intron 3  rs17878362  24  11  1  59  81.9  13  18.1  79.5  20.5  B  c.97<<<-29C>A>>>  intron 3  rs17883323  35  1  0  71  98.6  1  1.4  94.8  5.2  B  c.108G>A  p.Pro36Pro  exon 4  rs1800370  34  2  0  70  97.2  2  2.8  98.5  1.5  B  c",
            "29021639": "2.6  Pending  Benign   C3046T  1  2.6  Class 5  Pathogenic   V572L   1  2.6  NR  Uncertain   P218L  1  2.6  NR  NR   C6328T  3  7.9  Pending  Benign  <<< T10K>>>   1  2.6  NR  Uncertain   2929 delC  1  2.6  NR  Benign   3154 TC>AT  1  2.6  Pending  NR   C5972T  11  28.9  Class 1  Benign   T1011R (3260C>G)  1  |A1937G  1   G3031A  1  6076delGTTA  1   1163delTG  1  T289A  1   C5141T  1  C6448A  1   5637delG  1  C3045T  1   5154delTTTTC  1  P218L  1   E720X  1 <<< T10K>>>  1   N1742S  1  2929delC  1   2881delGACA  1  V572L  1   1499insA  1  C3260G  1   R1835X  1  6310delGA  1   2031delG  1  4772delA  1   K168E  1  A599|Pendiente  Benigna   C3046T  1  26  Clase 5   Patog\u00e9nica   V572L   1  2.6  NR  Incierta   P218L  1  2.6  NR  NR   C6328T  3  7.9  Pendiente  Benigna  <<< T10K>>>   1  2.6  NR  Incierta   2929 delC  1  2.6  NR  Benigna   3154 TC>AT  1  2.6  Pendiente  NR   C5972T  11  289  Class 1  Benigna   T1011R (3260C>G)  1|A1937G  1   G3031A  1  6076delGTTA  1   1163delTG  1  T289A  1   C5141T  1  C6448A  1   5637delG  1  C3045T  1   5154delTTTTC  1  P218L  1   E720X  1 <<< T10K>>>  1   N1742S  1  2929delC  1   2881delGACA  1  V572L  1   1499insA  1  C3260G  1   R1835X  1  6310delGA  1   2031delG  1  4772delA  1   K168E  1  A599|  0  0.0  Class 5  Pathogenic  W31X**  0  0.0  Pending  NR  C3046T  1  2.6  Class 5  Pathogenic  V572L  1  3034 delACAA**  T289A  C6448A P218L  C6328T<<< T10K>>>  2929 delC  3154 TC>AT  8  1 1 1 3 1 1 1  C5972T  11  6310 delGA  1  T1011R (3260C>G) 4772 delA A5996C  6062 insG  S1630X (5117C>G) N570S (1937A>G) T|ients  V1145F  C5141T  W1508X  A1708E  C5141T  W1712X  A1708E  T3014C  5356delT  23 1  A1937G  1  6076delGTTA  1  C3045T  1 1 1 1  T289A  C6448A P218L<<< T10K>>>  1  2929delC  1  6310delGA  2  C5117G  1 1 1 1  V572L  C3260G  4772delA  A5996C  2  6062insG  1  3154TC>AT  1 1 1 1 1 1 1  A1708E  2  5221delT  1  C5"
        },
        "chr13:g.32316490:A>G": {
            "27530658": "mitant actions of  the replication machinery and of a specialized reverse  Downloaded from Bioscientifica.com at 11/14/2018 08:42:19PM via free access<<< T10T>>>hematic Review A Fradet-Turcotte et\u00a0al."
        },
        "chr13:g.32316493:T>G": {
            "16762046": "/28 (10.7%)  N/D   Exon 1 (5' UTR)  1  c.-172G>T  -  16/33 (48.5%)  N/D   Intron 2  2  c.97+110A>G  -  3/46 (6.5%)  16/93 (17.2%)   Intron 3  4  c.226<<<-33T>G>>>  -  11/46 (23.9%)  23/93 (24.7%)   Intron 3  4  c.226-70T>A  -  20/46 (43.5%)  43/93 (46.2%)   Intron 3  4  c.226-72delA  -  2/46 (4.3%)  8/93 (8.6%)|3/93 (24.7%)   Intron 3  4  c.226-70T>A  -  20/46 (43.5%)  43/93 (46.2%)   Intron 3  4  c.226-72delA  -  2/46 (4.3%)  8/93 (8.6%)   Intron 3  4  c.226<<<-33T>G>>> and c.226-70T>A  -  4/46 (8.7%)  8/93 (8.6%)   Intron 3  4  c.226-33T>G and c.226-72delA  -  1/46 (2.1%)  0/93 (0%)   Intron 3  4  c.226-70T>A and c.|   Intron 3  4  c.226-72delA  -  2/46 (4.3%)  8/93 (8.6%)   Intron 3  4  c.226-33T>G and c.226-70T>A  -  4/46 (8.7%)  8/93 (8.6%)   Intron 3  4  c.226<<<-33T>G>>> and c.226-72delA  -  1/46 (2.1%)  0/93 (0%)   Intron 3  4  c.226-70T>A and c.226-72delA  -  1/46 (2.1%)  6/93 (6.5%)   Intron 4  5  c.344-36T>G  -  3|  3/28 (10.7%)  N/D  Exon 1 (5' UTR)  1  c.-172G>T  -  16/33 (48.5%)  N/D  Intron 2  2  c.97+110A>G  -  3/46 (6.5%)  16/93 (17.2%)  Intron 3  4  c.226<<<-33T>G>>>  -  11/46 (23.9%)  23/93 (24.7%)  Intron 3  4  c.226-70T>A  -  20/46 (43.5%)  43/93 (46.2%)  Intron 3  4  c.226-72delA  -  2/46 (4.3%)  8/93 (8.6%)  |  23/93 (24.7%)  Intron 3  4  c.226-70T>A  -  20/46 (43.5%)  43/93 (46.2%)  Intron 3  4  c.226-72delA  -  2/46 (4.3%)  8/93 (8.6%)  Intron 3  4  c.226<<<-33T>G>>> and c.226-70T>A  -  4/46 (8.7%)  8/93 (8.6%)  Intron 3  4  c.226-33T>G and c.226-72delA  -  1/46 (2.1%)  0/93 (0%)  Intron 3  4  c.226-70T>A and c.22|2%)  Intron 3  4  c.226-72delA  -  2/46 (4.3%)  8/93 (8.6%)  Intron 3  4  c.226-33T>G and c.226-70T>A  -  4/46 (8.7%)  8/93 (8.6%)  Intron 3  4  c.226<<<-33T>G>>> and c.226-72delA  -  1/46 (2.1%)  0/93 (0%)  Intron 3  4  c.226-70T>A and c.226-72delA  -  1/46 (2.1%)  6/93 (6.5%)  Intron 4  5  c.344-36T>G  -  3/4"
        },
```

**Comment by zfisch** [comment]
@rcurrie Nice! To your questions:

> Stick with '|', another character, or []

Seems like a good use case for a postgres array, but I imagine it'll work well either way. If we use concatenation instead, let's make sure we're not using characters found in the snippets themselves (e.g. do a search to make sure our delimiter isn't present anywhere). @falquaddoomi any strong opinions on this?

> PMID key in variants is a string although its an int - this is due to python serialization. The pmid field in the article dictionary is an int. Just a note and question if you want me to try and change it.

also not a big deal imo, though if it's easy to make them consistent may as well.

**Comment by rcurrie** [comment]
'|' does appear to show here and there. @zfisch is it easy to import the array? Have we used this feature of postgres before? If so then I'll switch to array. Otherwise propose using '|||' as a separator to sort of match <<<>>>.

**Comment by zfisch** [comment]
Yea, it's really straightforward, though we haven't used it before in this app. If we use arrays, here's a proposed json format per your previous sample:
```
{
    "papers": {
        "29298688": {
            "..."
        },
        "15564800": {
            "..."
        }
    },
    "variants": {
        "chr13:g.32316489:C>A": {
            "29298688": ["1", "2", "3"],
            "15564800": ["1", "2"]
        },
        "chr13:g.32316490:A>G": {
            "15564800": ["1"]
        },
        "chr13:g.32316493:T>G": {
            "29298688": ["2"]
        }
    }
}
```

**Comment by rcurrie** [comment]
Ok, I think I have it wrangled under control. Took quite a bit longer then I anticipated as I also had to deal with some of the mentions the crawler spits out with > 100 separate pieces slit by |. I've now switched to limiting them to only the first 3 as well as added a bit more pruning, de-duping etc...gory details [here](https://github.com/BRCAChallenge/pubMunch-BRCA/blob/docker/docker/export.ipynb). New export matching the above:

http://public.gi.ucsc.edu/~rcurrie/literature-05-12-2018-v1.json

48941 snippets now 

Example
```
literature["variants"]["chr13:g.32316470:G>T"]

{'17453335': ['Two novel BRCA2 germline mutations <<<(G4X>>> and 3783del10) are reported here for the first time.',
  'Each of the five BRCA2 mutations is observed only once; three are frameshift (3783del10, 4637delTA, 5950delCT), one is nonsense <<<(G4X>>>), and one is a single base substitution that results in aberrant splicing (7235G\xa0>\xa0A, skipping of exons 12–13) [15].',
  'h.gov/ Intramural_research/Lab_transfer/Bic/Member/ BRCA1_mutation_database.html). The two novel BRCA2 mutations described here for the first time are<<< G4X>>> and 3783del10, located in exons 2 and 11, respectively.',
  'Each of the five BRCA2 mutations is observed only once; three are frameshift (3783del10, 4637delTA, 5950delCT), one is nonsense <<<(G4X>>>), and one is a single base substitution that results in aberrant splicing (7235G > A, skipping of exons 12–13) [15].',
  'The two novel BRCA2 mutations described here for the first time are<<< G4X>>> and 3783del10, located in exons 2 and 11, respectively.'],
 '24010542': ['_5797delTA  11B  c.4409_4410delTA c.4415_4418delAGAA c.4767delA  11A  10  c.3555_3564del10 c.3919delG  9  p.Ser780X  7  c.2808_2811delAAAC c.2830delA <<< p.Gly4X>>>  5  21 23 20 22  8  c.1796_1800del5  3  1  6  c.658_659delGT  4  2  Fig. 1.'],
 '24312913': ['G, 461delTC [41]  4817A>G, 8477delAGA [41]  —   Greek (South eastern Europe)  5382insC and G1738R   [92]   G5331A, 3819delGTAAA [180]   8984delG [35],<<< G4X>>> and 3783del10 [92]  5382insC and G1738R   [92], 3782del10, 4512insT [180]   Slovenia  c.1193C>A [86]   c.181T>G, c.1687C>T, 5382insC,   c.844_850dupT',
  '8142del5 [28], exons 5–14 del Exons 1–17 del 5382insC [175]  c.6589delA [178, 179]  c.6589delA [178, 179]  4817A>G, 8477delAGA [41]  —  8984delG [35],<<< G4X>>> and 3783del10 [92]  5382insC and G1738R [92], 3782del10, 4512insT [180]  c.5101C>T, c.5433 5436delGGAA [86] c.7806−2A>G, c.5291C>G, c.3975 3978dupTGC'],
 '27225819': ['  95.96%   3.239   447.681   98.46%   BRCA2+   c.1447_1450ins4; p.Ala483AlaValfs   Path   #CQ3 TS   377.2   42.18%   227   120.384   57.25%   BRCA2+  <<< c.10G>T>>>; p.Gly4Ter   Path   #CQ5 TS   811.4   66.65%   400   173.040   89.41%   BRCA2+   c.5909C>A; p.Ser1970Ter   Path   #CQ7 TS   319.23   74.47%   105   3',
  '  3.239   447.681   98.46%   BRCA2+   c.1447_1450ins4; p.Ala483AlaValfs   Path   #CQ3 TS   377.2   42.18%   227   120.384   57.25%   BRCA2+   c.10G>T;<<< p.Gly4Ter>>>   Path   #CQ5 TS   811.4   66.65%   400   173.040   89.41%   BRCA2+   c.5909C>A; p.Ser1970Ter   Path   #CQ7 TS   319.23   74.47%   105   32.248   10.',
  'lT; p.Phe829Leufs c.1175_1214del40; p.Leu392Glnfs c.3756_3759delGTCT; p.Ser1253Argfs c.8764_8764delA; p.Ser2922Valfs c.1447_1450ins4; p.Ala483AlaValfs<<< c.10G>T>>>; p.Gly4Ter c.5909C>A; p.Ser1970Ter c.7617 1 1G>T c.6059_6062del4; p.Glu2020Valfs c.7251_7252delCA; p.His2417Glnfs c.9676delT; p.Tyr3226Ilefs c.4183C>T; p.Gln1395Ter c.5095C>T; p.Arg1699Trp  TABLE 3.',
  '829Leufs c.1175_1214del40; p.Leu392Glnfs c.3756_3759delGTCT; p.Ser1253Argfs c.8764_8764delA; p.Ser2922Valfs c.1447_1450ins4; p.Ala483AlaValfs c.10G>T;<<< p.Gly4Ter>>> c.5909C>A; p.Ser1970Ter c.7617 1 1G>T c.6059_6062del4; p.Glu2020Valfs c.7251_7252delCA; p.His2417Glnfs c.9676delT; p.Tyr3226Ilefs c.4183C>T; p.Gln1395Ter c.5095C>T; p.Arg1699Trp  TABLE 3.'],
 '28751759': ['65Leu) VUS, which was also present in this sample.  Laboratory 16, using an HRM pre-screening strategy, failed to identify the BRCA2 nonsense mutation<<< c.10G>T>>>; p.(Gly4*) in one sample.',
  '2Glnfs*5) NA 40 bp deletion (frameshift) 2 BRCA1 (NM_007294.3); BRCA2 (NM_000059.3) NA NA Control sample–no pathogenic mutations 3 BRCA2 (NM_000059.3)<<< c.10G4T>>>; p.(Gly4*) NA Point mutation (nonsense) 4 BRCA1 (NM_007294.3) c.442-?_547þ ?del, NA LGR: deletion (exon 8) 5 BRCA2 (NM_000059.3) c.5909C4A; p.(Ser197',
  '465Leu) VUS, which was also present in this sample. Laboratory 16, using an HRM pre-screening strategy, failed to identify the BRCA2 nonsense mutation<<< c.10G4T>>>; p.(Gly4*) in one sample.']}
```

**Comment by rcurrie** [comment]
@falquaddoomi Bit more detail on this new format and wrangling. Previously the 'snippets' sometimes has multiple strings separated by '|'. Most have a hand full but some have up to 100+ - all from the same article and and referenced against the same variant. I've pruned all these now splitting them up, keeping only 3 (based on the discussion on this weeks call) per variant per paper. These are all put into a single list per paper now so you don't need to do any splitting based on | - just possibly limiting if there is a bunch from the same paper.

**Comment by rcurrie** [comment]
@amycoffin Here's some preliminary stats etc...

Max mentions for a single variant: chr17:g.43106487:A>C 1411
Max mentions for a single paper from one variant: ('chr13:g.32319082:G>A', '27490902') 19
Paper referenced from the most variants: 21990134 1633

Stats for mentions by variant:
count    3754.000000
mean       13.037027
std        39.876334
min         1.000000
25%         2.000000
50%         4.000000
75%        11.000000
max      1411.000000

**Comment by zfisch** [comment]
@rcurrie is http://public.gi.ucsc.edu/~rcurrie/literature-05-12-2018-v1.json ready to be deployed to a dev machine or are there further changes underway?

**Comment by rcurrie** [comment]
@zfisch yes, please deploy. If the format works then we can freeze that so deploying will be a good test. I've been going through the new wrangling and feel relatively confident in it. But I now need other eyes to do spot checks to see if the snippets match up to the variants in real life (beyond my ad hoc) so deploying would be useful.

**Comment by zfisch** [comment]
@rcurrie @falquaddoomi @amycoffin @melissacline Latest literature search data is now up at http://52.229.50.68:8008. This includes the latest changes to the UI (max 3 snippets/mentions per variant per paper, uses arrays for snippets/mentions instead of splitting on '|', etc.).

Please take a look to make sure all of your latest work is properly reflected, and note any issues here!

**Comment by rcurrie** [comment]
@zfisch @falquaddoomi Suhweeet....

Variant with the most mentions: [chr17:g.43106487:A>C](http://52.229.50.68:8008/variant/172753)

**Comment by zfisch** [comment]
@melissacline @amycoffin what are our final steps for having this beta ready? Are there any further reviews necessary or any missing features? Should we have @falquaddoomi add a little beta tag on the literature search tile with a brief explanation that this feature is still under development?

New data will be available for review by Monday at the latest (possibly EOD today) -- do we want to include this first version of literature search when we push to beta next week?

**Comment by zfisch** [comment]
Ah, I see issue #947 -- is that our final list of req's before beta deploy?

**Comment by amycoffin** [comment]
@zfisch i agree that we should add a beta tag, mentioned it in issue #947. Once the three changes in that issue are made, I think its ready for beta

**Comment by melissacline** [comment]
For beta, let's add the note that Amy describes about contacting us about false positives / missing papers.  For now, we can use the standard "contact us" email, brca-exchange-contact@genomicsandhealth.org.  As a next step in refinement, we can link to a comment form, as Amy describes.  In the worst case, if I get lots of email, that will tell me that lots of people are interested in the feature!

As well as adding the beta badge, I feel like we should have a couple words to say what this feature is.  How about, in the heading of the table, in the column at the left (where the one at the right says "Citation Info"), say something like "Papers that appear to relate to this variant".?

**Comment by rcurrie** [comment]
I've started a new issue, #952 to track updating literature.json with papers added in the last month.

**Comment by rcurrie** [comment]
I've started an issue to track manual QC on randomly selected papers that fail to generate any mentiones https://github.com/BRCAChallenge/pubMunch-BRCA/issues/8

**Comment by rcurrie** [comment]
@melissacline your stats:
Crawled completed Nov 14, 2018
(embodied in literature-05-12-2018-v1.json which is our beta release of lit search):

```
Attempted to download 16980 PMIDs with BRCA in the title or abstract

Failed to download 2753(16%) of 14088 papers.
Reasons:
OK                     14088
invalidPdf               978
noCrawlerSuccess         817
httpError                302
noLicense                301
HighwirePdfNotValid      254
invalidHostname           55
pageErrorMessage          28
no_meta                    9
noOutlinkOrDoi             6
BeautifulSoupError         1
HtmlParseError             1
tooManySupplFiles          1

20934 variants in the BRCA Exchange database

Mentions/Snippets:
count    3754.000000

Per Paper:
mean       13.037027
std        39.876334
min         1.000000
25%         2.000000
50%         4.000000
75%        11.000000
max      1411.000000

Mentions stats per variant:
count    3754.000000
mean       13.037027
std        39.876334
min         1.000000
25%         2.000000
50%         4.000000
75%        11.000000
max      1411.000000
Name: pyhgvs_Genomic_Coordinate_38, dtype: float64

Variant with the most mentions:
chr17:g.43106487:A>C

Paper mentions stats:
count    2145.000000
mean       22.816317
std        58.991191
min         1.000000
25%         3.000000
50%         6.000000
75%        19.000000
max      1633.000000
Name: pmid, dtype: float64

Paper with the most mentions:
21990134

**Comment by rcurrie** [comment]
Done! Congratulations to the the vast team that contributed to this feature going way back to Max's original pubMunch. I just happened to show up at the 11th hour to add a few bits and bobs.
